lib/grammers-client/src/client/auth.rs

```diff
@@ -251,20 +281,128 @@ impl Client {
         {
             Ok(tl::enums::auth::Authorization::Authorization(x)) => {
                 // Safe to unwrap, Telegram won't send `UserEmpty` here.
-                Ok(x.user.try_into().unwrap())
+                Ok(SignInResult::SignedIn(x.user.try_into().unwrap()))
             }
             Ok(tl::enums::auth::Authorization::SignUpRequired(x)) => {
-                Err(SignInError::SignUpRequired {
+                Ok(SignInResult::SignUpRequired {
                     terms_of_service: x.terms_of_service.map(|tos| tos.into()),
                 })
             }
+            Err(InvocationError::Rpc(RpcError { name, .. }))
+                if name == "SESSION_PASSWORD_NEEDED" =>
+            {
+                let password_information = self.get_password_information().await?;
+                Ok(SignInResult::PasswordRequired {
+                    password_information,
+                })
+            }
             Err(InvocationError::Rpc(RpcError { name, .. })) if name.starts_with("PHONE_CODE_") => {
                 Err(SignInError::InvalidCode)
             }
             Err(error) => Err(SignInError::Other(error)),
         }
     }
 
+    /// Extract information needed for the two-factor authentication
+    /// It's called automatically when we get SESSION_PASSWORD_NEEDED error during sign in.
+    pub async fn get_password_information(
+        &mut self,
+    ) -> Result<tl::types::account::Password, SignInError> {
+        let request = tl::functions::account::GetPassword {};
+
+        let data: tl::enums::account::Password = match self.invoke(&request).await {
+            Ok(x) => x.into(),
+            Err(error) => return Err(SignInError::Other(error)),
+        };
+
+        let password_information: tl::types::account::Password = match data {
+            tl::enums::account::Password::Password(pass) => pass,
+        };
+
+        Ok(password_information)
```

Comment:
After changing the return error, this should probably be:

```suggestion
        let password = self.invoke(&request).await?.into();
        Ok(password)
```